setup.py -> pyproject.toml

ldoktor opened the pull request:
Modernize the setup.py -> pyproject.toml.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Adds aexpect_helper command-line entry point.

* **Chores**
  * Migrates to pyproject.toml-based packaging; removes legacy setup.py.
  * Prunes source manifests and adds global-exclude rules.
  * Enhances Makefile with develop, check, test, and pypi targets; modernizes install/build flows.
  * CI: runs an additional install step before self-tests.

* **Refactor**
  * Consolidates and reorganizes imports across modules without behavior changes.

* **Tests**
  * Minor test import reorderings; no behavioral changes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ldoktor committed:
pyproject.toml: Move away from setup.py

Use pyproject.toml to configure our project. This required to move the
aexpect_helper and a few updates to Makefile to avoid using setup.py.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
aexpect.scripts: Automated formatting

no functional changes, just black formatter.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
Style: reshuffle imports using isort

no changes, just reshuffling via isort.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
pyproject: Configure linters in pyproject

rather then specifying it in Makefile.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
Makefile: Use $(PROJECT) rather than hardcoded values

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
Makefile: Use "mkdir -p" rather than if/else

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
Makefile: Group by purpose

reorder by purpose and simplify descriptions.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor committed:
Makefile: Add targets to .PHONY

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

ldoktor commented:
@PaulYuuu since you are the expert in this matter, could you please review my approach? Note I haven't tested the deb/rpm targets, only updated them to be +- compliant. If anyone uses them, please verify and update them if needed.

ldoktor committed:
.github: Install aexpect before tests

functional tests require the scripts to be deployed.

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

PaulYuuu commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+[tool.setuptools.packages.find]
+where = ["."]
+include = ["aexpect*"]
+exclude = ["debian*", "docs*", "tests*", ".github*"]
```
`tool.setuptools.packages.find` is to tell the build system what kinds of files should in the wheel, as `include` already defined, `exclude` is not need. I have tested and the result is expected.

```bash
$ grep -A 4 'find' pyproject.toml
[tool.setuptools.packages.find]
where = ["."]
include = ["aexpect*"]

[tool.setuptools_scm]

$ python3 -m build --wheel > /dev/null 2>&1
$ unzip -l dist/*.whl
Archive:  dist/aexpect-1.8.0.post11+gdd032d41c.d20250825-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
      949  08-25-2025 08:07   aexpect/__init__.py
    70090  08-25-2025 08:07   aexpect/client.py
     3583  08-25-2025 08:06   aexpect/exceptions.py
    18866  08-25-2025 08:06   aexpect/ops_linux.py
    20891  08-25-2025 08:07   aexpect/ops_windows.py
    51814  08-25-2025 08:07   aexpect/remote.py
    41164  08-25-2025 08:07   aexpect/remote_door.py
    21044  08-25-2025 08:07   aexpect/rss_client.py
     3541  08-25-2025 08:07   aexpect/shared.py
        0  08-25-2025 08:07   aexpect/scripts/__init__.py
     8198  08-25-2025 08:07   aexpect/scripts/aexpect_helper.py
        0  08-25-2025 08:06   aexpect/utils/__init__.py
     2491  08-25-2025 08:06   aexpect/utils/astring.py
     1433  08-25-2025 08:06   aexpect/utils/data_factory.py
     2420  08-25-2025 08:06   aexpect/utils/path.py
     3233  08-25-2025 08:07   aexpect/utils/process.py
     1420  08-25-2025 08:07   aexpect/utils/wait.py
    18082  08-25-2025 11:13   aexpect-1.8.0.post11+gdd032d41c.d20250825.dist-info/licenses/LICENSE
     6743  08-25-2025 11:13   aexpect-1.8.0.post11+gdd032d41c.d20250825.dist-info/METADATA
       91  08-25-2025 11:13   aexpect-1.8.0.post11+gdd032d41c.d20250825.dist-info/WHEEL
       71  08-25-2025 11:13   aexpect-1.8.0.post11+gdd032d41c.d20250825.dist-info/entry_points.txt
        8  08-25-2025 11:13   aexpect-1.8.0.post11+gdd032d41c.d20250825.dist-info/top_level.txt
     2006  08-25-2025 11:13   aexpect-1.8.0.post11+gdd032d41c.d20250825.dist-info/RECORD
---------                     -------
   278138                     23 files
```

PaulYuuu replied:
Not sure if `[tool.setuptools.packages.find]` only for wheel and `MANIFEST.in` for sdist. as the `exclude` will not introduce regression, I am okay to keep it here.

PaulYuuu commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+testpaths = ["tests"]
+
+[tool.isort]
+profile = "black"
```
The default line length of isort is 79, but after switching the profile to black, I am not sure if it will change to 88.

PaulYuuu replied:
Black does not have rules for sorting import IIRC. Well, I searched the isort doc, it will change the line length if using black profile. https://pycqa.github.io/isort/docs/configuration/profiles.html

Black is for code style, I don't think it will sort the import order. With the current configuration, there are no big problem, just only tiny gap I think.

PaulYuuu commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+
+[tool.black]
+line-length = 79
+target-version = ["py38"]
```
Do we need to consider multiple version like in black doc `target-version = ["py311", "py312", "py313"]`?

ldoktor replied:
I went with the oldest to be safe-ish, but I'm open to adding 313 to be even safer, what do you think? (tbh I'm not sure what's the result then, does it run 2 checks with each version or does it do something smarter?)

PaulYuuu commented on Makefile:
```diff
@@ -1,100 +1,91 @@
-BUILDIR=$(CURDIR)/debian/aexpect
 PROJECT=aexpect
-VERSION="1.8.0"
+VERSION=$(shell $(PYTHON) -c "import setuptools_scm; print(setuptools_scm.get_version())")
```
I found the simple command line:
```bash
$ python -m setuptools_scm
1.8.0.post11+gdd032d41c
```

ldoktor commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+aexpect_helper = "aexpect.scripts.aexpect_helper:main"
+
+[tool.setuptools.packages.find]
+where = ["."]
```
I struggled with this of a while and you are right, it's probably redundant in this version so let me remove it.

ldoktor commented on MANIFEST.in:
```diff
@@ -0,0 +1,5 @@
+prune debian
+prune .github
+prune tests
+prune docs
```
As part of this I moved away from tarring sources manually, do you think I should tweak the MANIFEST to include those? The were included before but I thought they are not really useful and unnecessary increase the package size. I'm open to suggestions as I don't know how other packages handle this.

ldoktor replied:
Actually my bad, I tried using `build` but there were some issues so I kept the git archive

PaulYuuu replied:
Indeed, this is an open topic, it based on how we define the source archive. I checked the https://github.com/psf/black/archive/refs/tags/25.1.0.tar.gz, it includes all of files, but I don't think something like `.github` is useful in source.

Maybe we can keep the current MANIFEST, when someone requires more files in the source archive, we can modify the MANIFEST.in to include them.

PaulYuuu commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+[build-system]
+requires = ["setuptools>=61.0", "setuptools-scm>=8"]
```
Agree, to release a new tag, we add a new git tag, then build aexpect, it should a clear format like `1.9.0`, for the main branch and local development, we can see how many commits are added after the latest tag, like `1.8.0.post11+gd671f35f0`, that is useful for debug.

ldoktor commented:
Changes:

* get VERSION directly from `python`
* license string
* removed default "where" from pyproject
* added py312 and py313 to black linter config

ldoktor committed:
Makefile: Little improvements suggested by coderabbit

Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

YvanY0 reviewed (commented):
Hello @ldoktor, with the current change, it seems LGTM. only 1 place I would suggest that the isort config seting `line_length = 79`, to avoid the `black` profile override it from `79` to `88`.
https://pycqa.github.io/isort/docs/configuration/options.html#line-length
https://pycqa.github.io/isort/docs/configuration/profiles.html

ldoktor commented:
Changes:

* added `line_length = 79` to isort

pevogam commented:
> @PaulYuuu since you are the expert in this matter, could you please review my approach? Note I haven't tested the deb/rpm targets, only updated them to be +- compliant. If anyone uses them, please verify and update them if needed.

Hi @ldoktor, the RPM builds were failing and needed updates I added in https://github.com/avocado-framework/aexpect/pull/165. @PaulYuuu let me know what you think too, I provided arguments and motivation for the changes in each commit message.
